### PR TITLE
Add simple TrayerWidget

### PR DESCRIPTION
**Pull request by preisi**
This PR adds a new TrayerWidget to incorporate the system tray
utility 'trayer' into barpyrus. As of now it supports all screen corners
as positions on the bar, any other position might need further work.

The TrayerWidget itself instantiates a TrayerWatch object to start the
trayer process and a new X connection. It is then used to passively
watch for any updates of the trayer regarding its width using already
existing mechanisms of barpyrus.

### CODE REVIEW

File: barpyrus/trayer.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+            'tint': '0x29b2e',
+        }
+        for key,val in args.items():
+            self.default_args[key] = val
+        for key,val in self.default_args.items():
+            command += ["--%s" % (key), str(val)]
+
+        super(TrayerWatch,self).__init__(command)
+        self.proc.stdin.close()
+        self.proc.stdout.close()
```
Comment (t-wissmann) [inline barpyrus/trayer.py:29]: Let us hope trayer does not get killed by sigpipe :-)

File: barpyrus/trayer.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
+#!/usr/bin/env python3
+
+import threading
```
Comment (t-wissmann) [inline barpyrus/trayer.py]: We don't need this import anymore

File: barpyrus/trayer.py (hunk 1 above)
Comment (preisi) [reply barpyrus/trayer.py:29]: Mhh, good point. I might have gone over the top there. However, I did not notice any issues so far (maybe some longterm tests will prove me wrong :) )

File: barpyrus/trayer.py (hunk 2 above)
Comment (preisi) [reply barpyrus/trayer.py]: And its gone.

**Comment by t-wissmann** [comment]
The downside of trayer is that it only works if the panel (and thus also the trayer) is properly in the edge of a physical monitor maybe we can move it manually via python-xlib or use stalonetray instead. But all this can be done in a future PR.

I suggest to merge this PR as it is and make some more addjustments in the future to make it more configurable.

File: barpyrus/trayer.py (hunk 1 above)
Comment (t-wissmann) [reply barpyrus/trayer.py:29]: Maybe they handle sigpipe or maybe they don't write to stdout (they probably don't have to)

File: barpyrus/trayer.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
+#!/usr/bin/env python3
+
+from barpyrus.widgets import Widget
+from barpyrus.core import EventInput
+from Xlib.display import Display, X
+
+class TrayerWatch(EventInput):
+    def __init__(self, args = { }):
```
Comment (The-Compiler) [inline barpyrus/trayer.py]: Mutable default arguments are tricky in Python - all instances of this class would have the same dictionary this way.

See https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments

Might not matter here as there probably is just one `TrayerWatch`, but I'd still set `args=None` here and guard it with an `if args is not None:` or so below.

File: barpyrus/trayer.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            'expand': 'true',
+            'SetDockType': 'true',
+            'transparent': 'true',
+            'alpha': '0',
+            'height': '16',
+            'margin': '0',
+            'tint': '0x29b2e',
+        }
+        for key,val in args.items():
+            self.default_args[key] = val
```
Comment (The-Compiler) [inline barpyrus/trayer.py]: You can just do `self.default_args.update(args)` instead.

File: barpyrus/trayer.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+        super(TrayerWatch,self).__init__(command)
+        self.proc.stdin.close()
+        self.proc.stdout.close()
+
+        # search for running trayer window
+        self.display = Display()
+        root = self.display.screen().root
+        self.trayer = self.find_tray_window(root)
+        assert (self.trayer != None), 'Panel not found!'
```
Comment (The-Compiler) [inline barpyrus/trayer.py]:
```suggestion
        assert self.trayer is not None, 'Panel not found!'
```

No need for the parens, and `None` is typically compared with `is`.

File: barpyrus/trayer.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+    def fileno(self):
+        return self.display.fileno()
+
+    def process(self):
+        self.watch_trayer_non_blocking()
+
+
+class TrayerWidget(Widget):
+    def __init__(self, args = { }):
```
Comment (The-Compiler) [inline barpyrus/trayer.py]: Same as above

**Review by The-Compiler — CHANGES_REQUESTED** [review]

File: barpyrus/trayer.py (hunk 3 above)
Comment (t-wissmann) [reply barpyrus/trayer.py]: I wasn't aware of this :-O, thanks! (And I should check my python code...)

Comment (preisi) [reply barpyrus/trayer.py]: Oh, that is indeed good to know :o Thanks a lot!

File: barpyrus/trayer.py (hunk 4 above)
Comment (preisi) [reply barpyrus/trayer.py]: [x]

File: barpyrus/trayer.py (hunk 5 above)
Comment (preisi) [reply barpyrus/trayer.py]: [x]

**Pull request merged**